### PR TITLE
Dnsaddr support

### PR DESCRIPTION
**Pull request by igor-egorov**
Resolver for /dnsaddr records

The designed usage interface is AddressRepository::bootstrap.

Possible drawback - the callback passed to the bootstrap method is going to be called that many times as a number of peers discovered.

### CODE REVIEW

**Review by kamilsa — APPROVED** [review]

File: include/libp2p/peer/address_repository/inmem_address_repository.hpp [hunk 1]
```diff
@@ -22,8 +26,18 @@ namespace libp2p::peer {
   /**
    * @brief IN-memory implementation of Address repository.
    */
-  class InmemAddressRepository : public AddressRepository {
+  class InmemAddressRepository
+      : public AddressRepository,
+        std::enable_shared_from_this<InmemAddressRepository> {
```
[the remark this answers is not in the document]
Comment (igor-egorov) [reply include/libp2p/peer/address_repository/inmem_address_repository.hpp:31]: You are right. It should have `public` modifier stated explicitly

File: src/network/cares/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#
+# Copyright Soramitsu Co., Ltd. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+
+find_package(Threads)
```
[the remark this answers is not in the document]
Comment (igor-egorov) [reply src/network/cares/CMakeLists.txt:6]: True

File: src/network/cares/cares.cpp [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+    try {
+      std::thread worker([channel, request] {
+        ::ares_query(channel, request->uri.c_str(), ns_c_in, ns_t_txt,
+                     Ares::txtCallback, request.get());
+        Ares::waitAresChannel(channel);
+        ::ares_destroy(channel);
+      });
+      worker.detach();
+    } catch (std::runtime_error &e) {
```
Comment (xDimon) [inline src/network/cares/cares.cpp:123]: const-ref, please

File: src/network/cares/cares.cpp [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    auto processError = [request_ptr](int status_code) -> bool {
+      if (ARES_SUCCESS != status_code) {
+        reportError(request_ptr->io_context, request_ptr->callback,
+                    kQueryErrors.at(status_code));
+        removeRequest(request_ptr);
+        return true;
+      }
+      return false;
+    };
+    if (processError(status)) {
```
Comment (xDimon) [inline src/network/cares/cares.cpp:151]: It's not clear from the code how it works. Suggest to check status explicitly before error processing.

File: include/libp2p/network/cares/cares.hpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+#include <ares.h>
+#include <arpa/nameser.h>
+#include <sys/select.h>
+#include <boost/asio.hpp>
+#include <boost/optional.hpp>
+#include <libp2p/common/logger.hpp>
+#include <libp2p/outcome/outcome.hpp>
+
+namespace libp2p::network::cares {
```
Comment (xDimon) [inline include/libp2p/network/cares/cares.hpp:25]: This name confuses me. Think about to rename it, i.e. to **c_ares**

**Review by xDimon — CHANGES_REQUESTED** [review]

File: src/network/cares/cares.cpp [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    bool first_init = initialized_.compare_exchange_strong(expected, true);
+    if (not first_init) {
+      // atomic change failed, thus it was already initialized
+      log_->debug("C-ares library got initialized more than once");
+    } else if (auto status = ::ares_library_init(ARES_LIB_INIT_ALL);
+               ARES_SUCCESS != status) {
+      log_->error("Unable to initialize c-ares library - {}",
+                  ::ares_strerror(status));
+      // on library initialization failure we set initialized = false
+      assert(initialized_.compare_exchange_strong(expected, false));
```
[the remark this answers is not in the document]
Comment (igor-egorov) [reply src/network/cares/cares.cpp:78]: Works as expected. I've checked - an additional if for calling ares_init is not needed - it's already there.

**Review by xDimon — APPROVED** [review]
Good job!

**Pull request merged**